Lib/datetime.py

```diff
@@ -8,6 +8,31 @@
 import math as _math
 import sys
 
+
+class IsoCalendarDate(tuple):
```

Comment:
I think the name of this symbol should be `_IsoCalendarDate`, because it turns out that `datetime` doesn't have an `__all__`, which would make `IsoCalendarDate` presumptively public. We want the repr to continue to say `IsoCalendarDate`, though, so I see three options:

1. Declare the class name as `IsoCalendarDate` then rename the symbol and delete the old symbol, like so:

```python
class IsoCalendarDate(tuple):
    ...

# The class name shouldn't start with an underscore, but the symbol should
_IsoCalendarDate = IsoCalendarDate
del IsoCalendarDate
```

2. Set `__name__` and `__qualname__`, like so:

```python
class _IsoCalendarDate(tuple):
    ...
_IsoCalendarDate.__name__ = "IsoCalendarDate"
_IsoCalendarDate.__qualname__ = "IsoCalendarDate"
```

3. Change the `__repr__` implementation to hard-code the class name, like so:

```python
    def __repr__(self):
        return ('datetime.IsoCalendarDate'
            f'(year={self[0]}, week={self[1]}, weekday={self[2]})')
```

There are positives and negatives from each approach and, ultimately, it doesn't really matter much. I think the first approach is the most natural.

Reply:
Oh, I should also note, this is not a great place in the file for this class's definition. It's at the top, way before all the other class definitions. I would define it [right below the definition for `date`](https://github.com/python/cpython/blob/3f4db4a0bab073b768fae958e93288bd5d24eadd/Lib/datetime.py#L1140).

Reply:
@pganssle 
This makes `python attribute lookup IsoCalendarDate on datetime failed` whem pickling dump.